blog/_src/posts/2017-11-16-monotonicity-types-towards-a-type-system-for-eventual-consistency.md

```diff
@@ -0,0 +1,97 @@
+
+We'll leave you with an aspirational example, which demonstrates the need for a type system, rather than a more monolithic form of analysis, for proving functions monotone. This example is inspired by Lasp \[[Meiklejohn et al. (2015)](#ref1)\], where CRDTs are composed into programs through the use of monotone functions. It's a monotone function which takes two GCounter \[[Shapiro et al. (2011)](#ref3)\] values as arguments, and produces a GCounter maintaining the sum of its two arguments as an output.
+
+```
+getAt :: (m : NatMap, k : Nat) ⇒ Nat[↑ m, ? k]
+joinAt :: (m : NatMap, k : Nat, n : Nat) ⇒ NatMap[↑ m, ? k, ↑ n]
+span :: (x:NatMap) ⇒ Nat[↑ x]
+max :: (a : Nat, b : Nat) ⇒ Nat[↑ a, ↑ b]
+emptyMap :: NatMap
++ :: (x:Nat, y:Nat) ⇒ Nat[↑ x, ↑ y]
+> :: (x:Nat, y:Nat) ⇒ Bool[↑ x, ↓ y]
+
+type GCounter = { map : NatMap }
+
+sfun sumCounters(x : GCounter, y : GCounter) 
+ : GCounter[↑ x, ↑ y] =
+ let xMap : NatMap[↑ x, ↑ y] = x.map
+ let yMap : NatMap[↑ x, ↑ y] = y.map
+ let maxSpan : Nat[↑ x, ↑ y] = max (span xMap) (span yMap)
+ fun sumCell(k : Nat, acc : NatMap[↑ x, ↑ y]) 
+  : NatMap[↑ x, ↑ y] =
+  let cond : Bool[↑ x, ↓ y] = k > maxSpan
+   if cond then
+    acc
+   else
+    let acc' = joinAt acc k ((getAt xMap k) + (getAt yMap k))
+    sumCell (k+1) acc'
+ let initMap : IntArray[↑ x, ↑ y] = emptyMap
+ GCounter { array = sumCell 0 initMap }
+```
+
+The \\(\uparrow\\) symbol appearing in types is used to represent monotonic dependency on an sfun's argument. \\(\downarrow\\) is for antitonic dependency, and \\(?\\) is for unknown dependency. Sfun types, which may have multiple arguments, are written with \\(\Rightarrow\\) rather than \\(\to\\).
+
+While our system can handle much of this example, it can't handle everything yet, for several reasons. First, it involves an if condition which depends on the arguments of the enclosing sfun. To handle this, we would need to incorporate the notion of domain restriction into lifted reduction. Second, it involves recursion. This is problematic for us, because our system utilizes the fact that all well-typed programs terminate. We could partially address this by adding terminating fixpoint combinators, which allow recursion given some well-founded termination metric, as in \[[Vazou et al. (2014)](#ref5)\]. However, that would not be adequate for this particular function, which could require arbitrarily many levels of recursion depending on which values are supplied as arguments.
+
+So there's still much to do! If you're interested in more details behind the type system, have a look at Kevin's blog article, [Monotonicity Through Types](https://kevinclancy.github.io/2017/11/09/monotonicity-through-types.html), or have a look at the full [Monotonicity Types](https://infoscience.epfl.ch/record/231867) preprint for more.
```

Comment:
I think you should also point to Datafun, "Here's a different approach to tracking monotonicity through a type system"

Reply:
Good point. I haven't dealt with this yet, but I'll consider it.